src/Workspaces/Core/Portable/Utilities/AsyncLazyInitializer.cs

```diff
@@ -0,0 +1,114 @@
+        }
+
+        /// <summary>
+        /// Initializes a target reference type by using a specified asynchronous function if it hasn't already been
+        /// initialized.
+        /// </summary>
+        /// <typeparam name="T">The reference type of the reference to be initialized.</typeparam>
+        /// <typeparam name="TState">The type of the state object to pass to the accessor and value factory.</typeparam>
+        /// <param name="targetAccessor">An accessor that provides a reference of type <typeparamref name="T"/> to initialize if it hasn't already been initialized.</param>
+        /// <param name="valueFactory">The function that is called to initialize the reference.</param>
+        /// <param name="state">The state object to pass to the accessor and value factory.</param>
+        /// <returns>The initialized value of type <typeparamref name="T"/>.</returns>
+        /// <exception cref="NullReferenceException">
+        /// <para>If <paramref name="targetAccessor"/> is <see langword="null"/>.</para>
+        /// <para>-or-</para>
+        /// <para>If the target value is not already initialized and <paramref name="valueFactory"/> is <see langword="null"/>.</para>
+        /// </exception>
+        /// <exception cref="InvalidOperationException"><paramref name="valueFactory"/> returned <see langword="null"/>.</exception>
+        public static ValueTask<T> EnsureInitializedAsync<T, TState>(Accessor<T, TState> targetAccessor, Func<TState, ValueTask<T>> valueFactory, TState state)
+            where T : class
+        {
+            // Fast path
+            var value = Volatile.Read(ref targetAccessor(state));
+            if (!(value is null))
+            {
+                return new ValueTask<T>(value);
+            }
+
+            return EnsureInitializedCoreAsync(targetAccessor, valueFactory, state);
+        }
+
+        /// <summary>
+        /// Initializes a target reference type by using a specified asynchronous function (slow path).
+        /// </summary>
+        /// <typeparam name="T">The reference type of the reference to be initialized.</typeparam>
+        /// <param name="targetAccessor">An accessor that provides a reference of type <typeparamref name="T"/> to initialize.</param>
+        /// <param name="valueFactory">The function that is called to initialize the reference.</param>
+        /// <returns>The initialized value of type <typeparamref name="T"/>.</returns>
+        private static async ValueTask<T> EnsureInitializedCoreAsync<T>(Accessor<T> targetAccessor, Func<ValueTask<T>> valueFactory)
+            where T : class
+        {
+            var value = await valueFactory().ConfigureAwait(false);
+            if (value is null)
+            {
+                throw new InvalidOperationException(WorkspacesResources.AsyncLazy_StaticInit_InvalidOperation);
+            }
+
+            return Interlocked.CompareExchange(ref targetAccessor(), value, null) ?? value;
```

Comment:
You're using optimistic locking (in the SQL terminology) here, which allows the `valueFactory` to run many times and race to the finish. That should be very clearly documented in your xml comments.

Reply:
This is the same behavior as `LazyInitializer`, but I'll update the documentation anyway.